python/ql/src/experimental/Security/CWE-285/PamAuthorization.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Authorization bypass due to incorrect usage of PAM
+ * @description Using only the `pam_authenticate` call to check the validity of a login can lead to a authorization bypass.
+ * @kind problem
+ * @problem.severity warning
+ * @id py/pam-auth-bypass
+ * @tags security
+ *       external/cwe/cwe-285
+ */
+
+import python
+import semmle.python.ApiGraphs
+import experimental.semmle.python.Concepts
+import semmle.python.dataflow.new.TaintTracking
+
+private class LibPam extends API::Node {
+  LibPam() {
+    exists(
+      API::Node cdll, API::Node find_library, API::Node libpam, API::CallNode cdll_call,
+      API::CallNode find_lib_call, StrConst str
+    |
+      API::moduleImport("ctypes").getMember("CDLL") = cdll and
+      find_library = API::moduleImport("ctypes.util").getMember("find_library") and
+      cdll_call = cdll.getACall() and
+      find_lib_call = find_library.getACall() and
+      DataFlow::localFlow(DataFlow::exprNode(str), find_lib_call.getArg(0)) and
+      str.getText() = "pam" and
+      cdll_call.getArg(0) = find_lib_call and
+      libpam = cdll_call.getReturn()
+    |
+      libpam = this
+    )
```

Comment:
You can reduce the number of variables here. For example, you can remove the `cdll` variable altogether by doing

```
cdll_call = API::moduleImport("ctypes").getMember("CDLL").getACall()
```

Notice that you should also be using camelCase for CodeQL code, so it should really be `cdllCall`